SofaKernel/modules/SofaCore/SofaCore_test/ObjectFactory_test.cpp

```diff
@@ -0,0 +1,123 @@
+* under the terms of the GNU Lesser General Public License as published by    *
+* the Free Software Foundation; either version 2.1 of the License, or (at     *
+* your option) any later version.                                             *
+*                                                                             *
+* This program is distributed in the hope that it will be useful, but WITHOUT *
+* ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or       *
+* FITNESS FOR A PARTICULAR PURPOSE. See the GNU Lesser General Public License *
+* for more details.                                                           *
+*                                                                             *
+* You should have received a copy of the GNU Lesser General Public License    *
+* along with this program. If not, see <http://www.gnu.org/licenses/>.        *
+*******************************************************************************
+* Authors: The SOFA Team and external contributors (see Authors.txt)          *
+*                                                                             *
+* Contact information: contact@sofa-framework.org                             *
+******************************************************************************/
+#include <sofa/core/ObjectFactory.h>
+using sofa::core::ObjectFactory;
+using sofa::core::RegisterObject;
+
+#include <sofa/testing/BaseTest.h>
+using sofa::testing::BaseTest ;
+
+namespace
+{
+
+template<class Type>
+class TestObject : public sofa::core::objectmodel::BaseObject
+{
+public:
+    SOFA_CLASS(SOFA_TEMPLATE(TestObject, Type), sofa::core::objectmodel::BaseObject);
+};
+
+template<class Type>
+class TestObject2 : public sofa::core::objectmodel::BaseObject
+{
+public:
+    SOFA_CLASS(SOFA_TEMPLATE(TestObject2, Type), sofa::core::objectmodel::BaseObject);
+};
+
+int A = RegisterObject("Loads a plugin and import its content into the current namespace. ")
+        .add< TestObject<int> >();
+int B1 = RegisterObject("Loads a plugin and import its content into the current namespace. ")
+        .add< TestObject<double> >();
+int B2 = RegisterObject("Loads a plugin and import its content into the current namespace. ")
+        .add< TestObject<long> >();
+int B3 = RegisterObject("Loads a plugin and import its content into the current namespace. ")
+        .add< TestObject2<int> >();
+
+class ObjectFactory_test: public BaseTest
+{
+public:
+    void testDuplicatedRegistration()
+    {
+        {
+            EXPECT_MSG_EMIT(Warning);
+            int C = RegisterObject("Loads a plugin and import its content into the current namespace. ")
+                    .add< TestObject<long> >();
+            SOFA_UNUSED(C);
+        }
+    }
+
+    void testValidAlias()
+    {
+        ASSERT_TRUE(ObjectFactory::getInstance()->addAlias("FirstAlias", "TestObject"));
+    }
+
+    void testInvalidAlias()
+    {
+        {
```

Comment:
Why the double brackets?

Reply:
Because of EXPECT_MSG_EMIT. The validity of the test is scoped by the brackets.